### PR TITLE
Base64 cleanups/v1

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6964

### CODE REVIEW

**Comment by catenacyber** [comment]
@inashivb let me know when this is ready for review as you have pushed commits recently ;-)

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20135

**Comment by inashivb** [comment]
@catenacyber ready now :) thanks a lot for taking time!

File: src/util-base64.c [hunk 1]
```diff
@@ -185,16 +185,16 @@ Base64Ecode DecodeBase64(uint8_t *dest, uint32_t dest_size, const uint8_t *src,
...
         DecodeBase64Block(dptr, b64);
+        *decoded_bytes += numDecoded_blk;
         *consumed_bytes += bbidx;
     }
 
     /* Finish remaining b64 bytes by padding */
     if (valid && bbidx > 0 && (mode != BASE64_MODE_RFC2045)) {
         /* Decode remaining */
-        *decoded_bytes += ASCII_BLOCK - (B64_BLOCK - bbidx);
         DecodeBase64Block(dptr, b64);
```
Comment (catenacyber) [inline src/util-base64.c]: Why does this change matter ?

File: src/util-base64.c [hunk 2]
```diff
@@ -93,68 +94,136 @@ static inline void DecodeBase64Block(uint8_t ascii[ASCII_BLOCK], uint8_t b64[B64
     ascii[2] = (uint8_t) (b64[2] << 6) | (b64[3]);
 }
 
+static inline bool AmpleDestinationSize(
```
Comment (catenacyber) [inline src/util-base64.c:97]: Seems overkill to have such a simple function :-p

File: src/util-base64.c [hunk 3]
```diff
@@ -366,7 +366,7 @@ static int B64DecodeInCompleteStringWSp(void)
     const char *src = "SGVs bG8 gV29y bGQ";
     const char *fin_str = "Hello Wor";
     TEST_RFC2045(src, fin_str, strlen(fin_str) + 1 /* 12 B in dest_size */, strlen(fin_str),
-            strlen(src) - 3, BASE64_ECODE_OK);
+            strlen(src) - 4, BASE64_ECODE_OK);
```
Comment (catenacyber) [inline src/util-base64.c:369]: Why this change ?

File: src/util-base64.c [hunk 4]
```diff
@@ -93,68 +94,136 @@ static inline void DecodeBase64Block(uint8_t ascii[ASCII_BLOCK], uint8_t b64[B64
...
+ * \param consumed_bytes number of bytes successfully consumed by the decoder
+ * \param decoded_bytes number of bytes successfully decoded by the decoder
  *
- * \return Error code indicating success or failures with parsing
+ * \return Base64Ecode BASE64_ECODE_OK if all went well
+ *                     BASE64_ECODE_BUF if destination buffer got full before all could be decoded
  */
-Base64Ecode DecodeBase64(uint8_t *dest, uint32_t dest_size, const uint8_t *src, uint32_t len,
-        uint32_t *consumed_bytes, uint32_t *decoded_bytes, Base64Mode mode)
+static inline Base64Ecode DecodeBase64RFC2045(uint8_t *dest, uint32_t dest_size, const uint8_t *src,
```
Comment (catenacyber) [inline src/util-base64.c:125]: I am not sure about the philosophy of the commit as it seems to bring in more complexity... (multiply by 3 the lines of code to review...)

File: src/util-base64.c (hunk 1 above)
Comment (inashivb) [reply src/util-base64.c]: It doesn't anymore. Was trying to do small cleanups before it became too complicated and I ended up doing the work in this PR.
Will remove in the next rev.

File: src/util-base64.c (hunk 2 above)
Comment (inashivb) [reply src/util-base64.c:97]: Indeed. :P Was hoping to do more things at first here. Now, it's not much use. Will remove. :)

File: src/util-base64.c (hunk 3 above)
Comment (victorjulien) [reply src/util-base64.c:369]: In general, I think it would be nice to use explicit sizes in these tests. These strlen derived sizes make it harder to follow what is happening.

Comment (inashivb) [reply src/util-base64.c:369]: Because the initial consumed bytes were incorrect. See below.
Destination size: 10
B64 encoded string: SGVs bG8 gV29y bGQ
Mode: RFC 2045

We process them the following way.
First pass: `SGVs` -> `Hel`; consumed bytes: 4; decoded bytes: 3
Second pass: ` bG8 g` -> `bG8g` (skipped spaces as they're not base64 alphabet and any such chars must be skipped as per RFC 2045) -> `lo `, consumed bytes: 6 + 4; decoded bytes: 3 + 3
Third pass: `V29y` -> `Wor`; consumed bytes: 6 + 4 + 4; decoded bytes: 3 + 3 +3
Fourth pass: ` bGQ` but, there isn't enough data or destination size to decode, so we do not process it.

So, finally, consumed bytes: 6 + 4 + 4 = 14 i.e. strlen(src) - 4 where strlen(src) = 18.

File: src/util-base64.c (hunk 4 above)
Comment (inashivb) [reply src/util-base64.c:125]: ok. Have you tried opening the changes in place in the file? diff is messy indeed. Lmk if you'd still like to reject this. Thank you!

Comment (inashivb) [reply src/util-base64.c:125]: Also to note: I added lots of fn docs, that may be adding to the increase in the number of lines.
